README.md
# Accurate-core &middot; [![NPM version](https://img.shields.io/npm/v/accurate-core.svg?style=flat)](https://www.npmjs.com/package/accurate-core) [![NPM downloads](https://img.shields.io/npm/dm/accurate-core.svg?style=flat)](https://www.npmjs.com/package/accurate-core) [![Build Status](https://travis-ci.org/Ipxxiao/accurate-core.svg?branch=master)](https://travis-ci.org/Ipxxiao/accurate-core) [![Coverage Status](https://coveralls.io/repos/github/Ipxxiao/accurate-core/badge.svg?branch=master)](https://coveralls.io/github/Ipxxiao/accurate-core?branch=master) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

- An lightweight and high performance precision calculation for JavaScript and Node.js.
- 解决 JavaScript 与 Node.js 精度计算（浮点数计算精度）问题，轻量级、性能较高。

## Browsers support

| [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png" alt="IE / Edge" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br />IE / Edge | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png" alt="Firefox" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br />Firefox | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png" alt="Chrome" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br />Chrome | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png" alt="Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br />Safari | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari-ios/safari-ios_48x48.png" alt="iOS Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br />iOS Safari | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/opera/opera_48x48.png" alt="Opera" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br />Opera |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| IE8+ Edge+                                                                                                                                                                                                      | last 10 versions                                                                                                                                                                                                  | last 10 versions                                                                                                                                                                                              | last 2 versions                                                                                                                                                                                               | last 2 versions                                                                                                                                                                                                               | last 2 versions                                                                                                                                                                                           |

## Install

``` base
npm i accurate-core -S
or
yarn add accurate-core
```

## Features

- Lightweight.
- High performance.
- No dependencies.
- Runs on any JavaScript engine.
- Is easily extensible.
- Open source.

- 轻量级。
- 性能较高。
- 没有依赖。
- 运行在任何 JavaScript 引擎。
- 易于扩展。
- 开源的。

## More features

- See [accurate](https://github.com/Ipxxiao/accurate) for more features.
- 更多功能请查看 [accurate](https://github.com/Ipxxiao/accurate)。

## Usage

``` javascript 
import accurateCalc, {
  add, addition, subtract, subtraction, multiply, mul, multiplication, division, divide, modulo, modulus
} from 'accurate-core'
or
const accurateCalc, {
  add, addition, subtract, subtraction, multiply, mul, multiplication, division, divide, modulo, modulus
} = require('accurate-core')
```


. **add（加法）**
``` javascript
>
1.1 + 0.3
accurateCalc.add(1.1, 0.3)
addition(1.1, 0.3)

<-
1.4000000000000001
1.4
1.4
```

. **subtract（减法）**
``` javascript
>
1.1 - 0.2
accurateCalc.subtract(1.1, 0.2)
subtraction(1.1, 0.2)

<-
0.9000000000000001
0.9
0.9
```

. **multiply（乘法）**
``` javascript
>
1.1 * 0.1
accurateCalc.multiply(1.1, 0.1)
accurateCalc.mul(1.1, 0.1)
multiplication(1.1, 0.1)

<-
0.11000000000000001
0.11
0.11
0.11
```

. **division（除法）**
``` javascript
>
1.1 / 10
accurateCalc.division(1.1, 10)
accurateCalc.divide(1.1, 10)

<-
0.11000000000000001
0.11
0.11
```

. **modulo（取模）**
``` javascript
>
1.1 % 1
accurateCalc.modulo(1.1, 1)
accurateCalc.modulus(1.1, 1)

<-
0.10000000000000009
0.1
0.1
```

## More
- [MIT License](https://github.com/Ipxxiao/accurate-core/blob/master/LICENSE)
- [Changelog](https://github.com/Ipxxiao/accurate-core/blob/master/CHANGELOG.md)
- [Test Case](https://github.com/Ipxxiao/accurate-core/blob/master/__tests__/index.spec.ts)
- [Docs](https://github.com/Ipxxiao/accurate-core/tree/master/docs)


CHANGELOG.md
# Release history

**All notable changes to this project will be documented in this file.**

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<details>
  <summary><strong>Guiding Principles</strong></summary>

-   Changelogs are for humans, not machines.
-   There should be an entry for every single version.
-   The same types of changes should be grouped.
-   Versions and sections should be linkable.
-   The latest version comes first.
-   The release date of each versions is displayed.
-   Mention whether you follow Semantic Versioning.

</details>

<details>
  <summary><strong>Types of changes</strong></summary>

Changelog entries are classified using the following labels _(from [keep-a-changelog](http://keepachangelog.com/)_):

-   `Added` for new features.
-   `Changed` for changes in existing functionality.
-   `Deprecated` for soon-to-be removed features.
-   `Removed` for now removed features.
-   `Fixed` for any bug fixes.
-   `Security` in case of vulnerabilities.

</details>

## 1.0.6 (2022-02-23)

-   Security update of dependabot alerts

## 1.0.5 (2021-04-07)

-   Changed yarn.lock

## 1.0.4 (2021-04-01)

-   Security update jest version

## 1.0.3 (2020-12-28)

-   Security update jest version

## 1.0.2 (2020-07-31)

-   Remove Codecov of dependencies

## 1.0.1 (2020-07-24)

-   First release, from the [accurate](https://github.com/Ipxxiao/accurate) extraction

[keep-a-changelog]: https://github.com/olivierlacan/keep-a-changelog


__tests__/index.spec.ts
import accurateCalc, {
	add, addition, subtract, subtraction, multiply, mul, multiplication, division, divide, modulo, modulus
} from '../src/index'

const methods = ['add', 'subtract', 'multiply', 'division', 'modulo']

describe('accurateCalc', () => {
	methods.forEach(par => {
		it(`should have ${par} method`, () => {
			expect(accurateCalc).toHaveProperty(par)
			expect(accurateCalc[par]).toBeInstanceOf(Function)
		})
	})

	it(`add`, () => {
		expect(accurateCalc.add(1.1, 0.3)).toEqual(1.4)
	})

	it(`subtract`, () => {
		expect(accurateCalc.subtract(1.1, 0.2)).toEqual(0.9)
	})

	it(`multiply`, () => {
		expect(accurateCalc.multiply(1.1, 0.1)).toEqual(0.11)
	})

	it(`division`, () => {
		expect(accurateCalc.division(1.1, 10)).toEqual(0.11)
	})

	it(`modulo`, () => {
		expect(accurateCalc.modulo(1.1, 0.2)).toEqual(0.1)
	})
})

src/index.ts
/**
 * accurate-core module.
 * @module accurate-core
 * @see doc https://github.com/Ipxxiao/accurate-core/tree/master/docs
 */
type OffsetData = {
    int1: number
    int2: number
    digits: number
}

/**
 * 获取小数位数
 *
 * @param {number} num
 * @returns {number}
 */
const getDecimalDigits = (num: number): number => {
    const decimal: string = num.toString().split('.')[1];

    if (decimal) {
        return decimal.length;
    } else {
        return 0;
    }
}

/**
 * 获取偏移数据
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {OffsetData}
 */
const getOffsetData = (num1: number, num2: number): OffsetData => {
    num1 = Number(num1);
    num2 = Number(num2);

    const len1: number = getDecimalDigits(num1);
    const len2: number = getDecimalDigits(num2);
    const digits: number = 10 ** Math.max(len1, len2);

    return {
        digits,
        int1: Math.round(num1 * digits),
        int2: Math.round(num2 * digits),
    }
}

/**
 * 精度加法计算
 *
 * @example
 * ```js
 * add(1.1, 0.3)
 * //=> 1.4
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export const add = (num1: number, num2: number): number => {
    const {
        int1,
        int2,
        digits,
    } = getOffsetData(num1, num2);

    return (int1 + int2) / digits;
}
export const addition = add;

/**
 * 精度减法计算
 *
 * @example
 * ```js
 * subtract(1.1, 0.2)
 * //=> 0.9
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export const subtract = (num1: number, num2: number): number => {
    return add(num1, -num2);
}
export const subtraction = subtract;

/**
 * 精度乘法计算
 *
 * @example
 * ```js
 * multiply(1.1, 0.1)
 * //=> 0.11
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export const multiply = (num1: number, num2: number): number => {
    const {
        int1,
        int2,
        digits,
    } = getOffsetData(num1, num2);

    return int1 * int2 / digits ** 2;
}
export const mul = multiply;
export const multiplication = multiply;

/**
 * 精度除法计算
 *
 * @example
 * ```js
 * division(1.1, 10)
 * //=> 0.11
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export const division = (num1: number, num2: number): number => {
    const {
        int1,
        int2,
    } = getOffsetData(num1, num2);

    return int1 / int2;
}
export const divide = division;

/**
 * 精度取模计算
 *
 * @example
 * ```js
 * modulo(1.1, 1)
 * //=> 0.1
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export const modulo = (num1: number, num2: number): number => {
    const {
        int1,
        int2,
        digits,
    } = getOffsetData(num1, num2);

    return int1 % int2 / digits;
}
export const modulus = modulo;

export default {
    add,
    addition,
    subtract,
    subtraction,
    mul,
    multiply,
    multiplication,
    division,
    divide,
    modulo,
    modulus,
};

libs/es/index.d.ts
/**
 * 精度加法计算
 *
 * @example
 * ```js
 * add(1.1, 0.3)
 * //=> 1.4
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export declare const add: (num1: number, num2: number) => number;
export declare const addition: (num1: number, num2: number) => number;
/**
 * 精度减法计算
 *
 * @example
 * ```js
 * subtract(1.1, 0.2)
 * //=> 0.9
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export declare const subtract: (num1: number, num2: number) => number;
export declare const subtraction: (num1: number, num2: number) => number;
/**
 * 精度乘法计算
 *
 * @example
 * ```js
 * multiply(1.1, 0.1)
 * //=> 0.11
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export declare const multiply: (num1: number, num2: number) => number;
export declare const mul: (num1: number, num2: number) => number;
export declare const multiplication: (num1: number, num2: number) => number;
/**
 * 精度除法计算
 *
 * @example
 * ```js
 * division(1.1, 10)
 * //=> 0.11
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export declare const division: (num1: number, num2: number) => number;
export declare const divide: (num1: number, num2: number) => number;
/**
 * 精度取模计算
 *
 * @example
 * ```js
 * modulo(1.1, 1)
 * //=> 0.1
 * ```
 *
 * @param {number} num1
 * @param {number} num2
 * @returns {number}
 */
export declare const modulo: (num1: number, num2: number) => number;
export declare const modulus: (num1: number, num2: number) => number;
declare const _default: {
    add: (num1: number, num2: number) => number;
    addition: (num1: number, num2: number) => number;
    subtract: (num1: number, num2: number) => number;
    subtraction: (num1: number, num2: number) => number;
    mul: (num1: number, num2: number) => number;
    multiply: (num1: number, num2: number) => number;
    multiplication: (num1: number, num2: number) => number;
    division: (num1: number, num2: number) => number;
    divide: (num1: number, num2: number) => number;
    modulo: (num1: number, num2: number) => number;
    modulus: (num1: number, num2: number) => number;
};
export default _default;


libs/umd/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.modulus = exports.modulo = exports.divide = exports.division = exports.multiplication = exports.mul = exports.multiply = exports.subtraction = exports.subtract = exports.addition = exports.add = void 0;
    /**
     * 获取小数位数
     *
     * @param {number} num
     * @returns {number}
     */
    var getDecimalDigits = function (num) {
        var decimal = num.toString().split('.')[1];
        if (decimal) {
            return decimal.length;
        }
        else {
            return 0;
        }
    };
    /**
     * 获取偏移数据
     *
     * @param {number} num1
     * @param {number} num2
     * @returns {OffsetData}
     */
    var getOffsetData = function (num1, num2) {
        num1 = Number(num1);
        num2 = Number(num2);
        var len1 = getDecimalDigits(num1);
        var len2 = getDecimalDigits(num2);
        var digits = Math.pow(10, Math.max(len1, len2));
        return {
            digits: digits,
            int1: Math.round(num1 * digits),
            int2: Math.round(num2 * digits),
        };
    };
    /**
     * 精度加法计算
     *
     * @example
     * ```js
     * add(1.1, 0.3)
     * //=> 1.4
     * ```
     *
     * @param {number} num1
     * @param {number} num2
     * @returns {number}
     */
    exports.add = function (num1, num2) {
        var _a = getOffsetData(num1, num2), int1 = _a.int1, int2 = _a.int2, digits = _a.digits;
        return (int1 + int2) / digits;
    };
    exports.addition = exports.add;
    /**
     * 精度减法计算
     *
     * @example
     * ```js
     * subtract(1.1, 0.2)
     * //=> 0.9
     * ```
     *
     * @param {number} num1
     * @param {number} num2
     * @returns {number}
     */
    exports.subtract = function (num1, num2) {
        return exports.add(num1, -num2);
    };
    exports.subtraction = exports.subtract;
    /**
     * 精度乘法计算
     *
     * @example
     * ```js
     * multiply(1.1, 0.1)
     * //=> 0.11
     * ```
     *
     * @param {number} num1
     * @param {number} num2
     * @returns {number}
     */
    exports.multiply = function (num1, num2) {
        var _a = getOffsetData(num1, num2), int1 = _a.int1, int2 = _a.int2, digits = _a.digits;
        return int1 * int2 / Math.pow(digits, 2);
    };
    exports.mul = exports.multiply;
    exports.multiplication = exports.multiply;
    /**
     * 精度除法计算
     *
     * @example
     * ```js
     * division(1.1, 10)
     * //=> 0.11
     * ```
     *
     * @param {number} num1
     * @param {number} num2
     * @returns {number}
     */
    exports.division = function (num1, num2) {
        var _a = getOffsetData(num1, num2), int1 = _a.int1, int2 = _a.int2;
        return int1 / int2;
    };
    exports.divide = exports.division;
    /**
     * 精度取模计算
     *
     * @example
     * ```js
     * modulo(1.1, 1)
     * //=> 0.1
     * ```
     *
     * @param {number} num1
     * @param {number} num2
     * @returns {number}
     */
    exports.modulo = function (num1, num2) {
        var _a = getOffsetData(num1, num2), int1 = _a.int1, int2 = _a.int2, digits = _a.digits;
        return int1 % int2 / digits;
    };
    exports.modulus = exports.modulo;
    exports.default = {
        add: exports.add,
        addition: exports.addition,
        subtract: exports.subtract,
        subtraction: exports.subtraction,
        mul: exports.mul,
        multiply: exports.multiply,
        multiplication: exports.multiplication,
        division: exports.division,
        divide: exports.divide,
        modulo: exports.modulo,
        modulus: exports.modulus,
    };
});
//# sourceMappingURL=index.js.map